File: sdk/go/jwks.go
package certnode

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"
)

// JWKSManager manages JWKS fetching and caching.
type JWKSManager struct {
	ttl     time.Duration
	client  *http.Client
	mu      sync.RWMutex
	cache   *JWKS
	cached  time.Time
	fetcher func(url string) (*JWKS, error)
}

// JWKSManagerOptions configures the JWKSManager.
type JWKSManagerOptions struct {
	TTL     time.Duration
	Client  *http.Client
	Fetcher func(url string) (*JWKS, error)
}

// NewJWKSManager creates a new JWKS manager with caching.
func NewJWKSManager(opts *JWKSManagerOptions) *JWKSManager {
	if opts == nil {
		opts = &JWKSManagerOptions{}
	}

	ttl := opts.TTL
	if ttl == 0 {
		ttl = 5 * time.Minute // Default 5 minute cache
	}

	client := opts.Client
	if client == nil {
		client = &http.Client{
			Timeout: 30 * time.Second,
		}
	}

	manager := &JWKSManager{
		ttl:    ttl,
		client: client,
	}

	if opts.Fetcher != nil {
		manager.fetcher = opts.Fetcher
	} else {
		manager.fetcher = manager.defaultFetcher
	}

	return manager
}

// FetchFromURL fetches JWKS from URL with caching.
func (m *JWKSManager) FetchFromURL(url string) (*JWKS, error) {
	// Check cache first
	m.mu.RLock()
	if m.cache != nil && time.Since(m.cached) < m.ttl {
		jwks := m.cache
		m.mu.RUnlock()
		return jwks, nil
	}
	m.mu.RUnlock()

	// Fetch fresh JWKS
	jwks, err := m.fetcher(url)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch JWKS from %s: %w", url, err)
	}

	// Validate JWKS
	if err := m.ValidateJWKS(jwks); err != nil {
		return nil, fmt.Errorf("invalid JWKS from %s: %w", url, err)
	}

	// Update cache
	m.mu.Lock()
	m.cache = jwks
	m.cached = time.Now()
	m.mu.Unlock()

	return jwks, nil
}

// SetFromObject sets JWKS from object with validation.
func (m *JWKSManager) SetFromObject(jwks *JWKS) error {
	if err := m.ValidateJWKS(jwks); err != nil {
		return err
	}

	m.mu.Lock()
	m.cache = jwks
	m.cached = time.Now()
	m.mu.Unlock()

	return nil
}

// GetFresh returns cached JWKS if still fresh.
func (m *JWKSManager) GetFresh() *JWKS {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if m.cache != nil && time.Since(m.cached) < m.ttl {
		return m.cache
	}

	return nil
}

// Thumbprints returns thumbprints of all keys in JWKS.
func (m *JWKSManager) Thumbprints(jwks *JWKS) ([]string, error) {
	if jwks == nil {
		m.mu.RLock()
		jwks = m.cache
		m.mu.RUnlock()
	}

	if jwks == nil {
		return nil, fmt.Errorf("no JWKS available")
	}

	var thumbprints []string
	for i := range jwks.Keys {
		thumbprint, err := JWKThumbprint(&jwks.Keys[i])
		if err != nil {
			// Skip keys that can't generate thumbprints
			continue
		}
		thumbprints = append(thumbprints, thumbprint)
	}

	return thumbprints, nil
}

// ValidateJWKS validates a JWKS structure.
func (m *JWKSManager) ValidateJWKS(jwks *JWKS) error {
	if jwks == nil {
		return fmt.Errorf("JWKS is nil")
	}

	if len(jwks.Keys) == 0 {
		return fmt.Errorf("JWKS contains no keys")
	}

	// Validate each key
	for i, key := range jwks.Keys {
		if err := m.ValidateJWK(&key); err != nil {
			return fmt.Errorf("key %d: %w", i, err)
		}
	}

	return nil
}

// ValidateJWK validates a single JWK.
func (m *JWKSManager) ValidateJWK(jwk *JWK) error {
	if jwk.Kty == "" {
		return fmt.Errorf("missing kty field")
	}

	switch jwk.Kty {
	case "EC":
		if jwk.Crv != "P-256" {
			return fmt.Errorf("only P-256 curve supported for EC keys, got %s", jwk.Crv)
		}
		if jwk.X == "" || jwk.Y == "" {
			return fmt.Errorf("EC key missing x or y coordinate")
		}
	case "OKP":
		if jwk.Crv != "Ed25519" {
			return fmt.Errorf("only Ed25519 curve supported for OKP keys, got %s", jwk.Crv)
		}
		if jwk.X == "" {
			return fmt.Errorf("OKP key missing x coordinate")
		}
	default:
		return fmt.Errorf("unsupported key type: %s. Use 'EC' or 'OKP'", jwk.Kty)
	}

	return nil
}

// defaultFetcher is the default JWKS fetcher using HTTP.
func (m *JWKSManager) defaultFetcher(url string) (*JWKS, error) {
	resp, err := m.client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	var jwks JWKS
	if err := json.Unmarshal(body, &jwks); err != nil {
		return nil, fmt.Errorf("failed to parse JWKS JSON: %w", err)
	}

	return &jwks, nil
}

// FetchJWKS is a simple JWKS fetch function without caching.
func FetchJWKS(url string) (*JWKS, error) {
	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	resp, err := client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch JWKS: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d from %s", resp.StatusCode, url)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var jwks JWKS
	if err := json.Unmarshal(body, &jwks); err != nil {
		return nil, fmt.Errorf("failed to parse JWKS: %w", err)
	}

	// Basic validation
	if len(jwks.Keys) == 0 {
		return nil, fmt.Errorf("JWKS contains no keys")
	}

	return &jwks, nil
}

File: sdk/rust/src/jwks.rs
//! JWKS management for CertNode SDK.

use crate::{CertNodeError, Jwk, Jwks, Result};
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};

/// JWKS manager with caching and async support.
///
/// Provides automatic JWKS fetching and caching with configurable TTL.
/// Thread-safe and optimized for high-performance scenarios.
///
/// # Examples
///
/// ```rust,no_run
/// use certnode::JwksManager;
/// use std::time::Duration;
///
/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
/// let manager = JwksManager::new(Duration::from_secs(300)); // 5 minute cache
///
/// let jwks = manager.fetch_from_url("https://api.certnode.io/.well-known/jwks.json").await?;
/// println!("Fetched {} keys", jwks.keys.len());
/// # Ok(())
/// # }
/// ```
#[derive(Debug)]
pub struct JwksManager {
    ttl: Duration,
    cache: Arc<RwLock<Option<CachedJwks>>>,
    #[cfg(feature = "jwks-fetch")]
    client: reqwest::Client,
}

#[derive(Debug, Clone)]
struct CachedJwks {
    jwks: Jwks,
    cached_at: Instant,
}

impl JwksManager {
    /// Create a new JWKS manager with specified TTL.
    ///
    /// # Arguments
    ///
    /// * `ttl` - Time-to-live for cached JWKS
    pub fn new(ttl: Duration) -> Self {
        Self {
            ttl,
            cache: Arc::new(RwLock::new(None)),
            #[cfg(feature = "jwks-fetch")]
            client: reqwest::Client::new(),
        }
    }

    /// Create a new JWKS manager with custom HTTP client.
    #[cfg(feature = "jwks-fetch")]
    #[cfg_attr(docsrs, doc(cfg(feature = "jwks-fetch")))]
    pub fn with_client(ttl: Duration, client: reqwest::Client) -> Self {
        Self {
            ttl,
            cache: Arc::new(RwLock::new(None)),
            client,
        }
    }

    /// Fetch JWKS from URL with caching.
    ///
    /// Returns cached JWKS if still fresh, otherwise fetches from the URL.
    /// This method is thread-safe and can be called concurrently.
    ///
    /// # Arguments
    ///
    /// * `url` - URL to fetch JWKS from
    ///
    /// # Returns
    ///
    /// Returns the JWKS or an error if fetch/parse fails.
    #[cfg(feature = "jwks-fetch")]
    #[cfg_attr(docsrs, doc(cfg(feature = "jwks-fetch")))]
    pub async fn fetch_from_url(&self, url: &str) -> Result<Jwks> {
        // Check cache first (read lock)
        {
            let cache = self.cache.read().unwrap();
            if let Some(cached) = cache.as_ref() {
                if cached.cached_at.elapsed() < self.ttl {
                    return Ok(cached.jwks.clone());
                }
            }
        }

        // Fetch fresh JWKS
        let response = self
            .client
            .get(url)
            .timeout(Duration::from_secs(30))
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(CertNodeError::NetworkError(format!(
                "HTTP {} from {}",
                response.status(),
                url
            )));
        }

        let jwks: Jwks = response.json().await?;

        // Validate JWKS
        self.validate_jwks(&jwks)?;

        // Update cache (write lock)
        {
            let mut cache = self.cache.write().unwrap();
            *cache = Some(CachedJwks {
                jwks: jwks.clone(),
                cached_at: Instant::now(),
            });
        }

        Ok(jwks)
    }

    /// Set JWKS from object with validation.
    ///
    /// # Arguments
    ///
    /// * `jwks` - The JWKS to cache
    pub fn set_from_object(&self, jwks: Jwks) -> Result<()> {
        self.validate_jwks(&jwks)?;

        let mut cache = self.cache.write().unwrap();
        *cache = Some(CachedJwks {
            jwks,
            cached_at: Instant::now(),
        });

        Ok(())
    }

    /// Get cached JWKS if still fresh.
    ///
    /// Returns `None` if no JWKS is cached or if the cache has expired.
    pub fn get_fresh(&self) -> Option<Jwks> {
        let cache = self.cache.read().unwrap();
        cache.as_ref().and_then(|cached| {
            if cached.cached_at.elapsed() < self.ttl {
                Some(cached.jwks.clone())
            } else {
                None
            }
        })
    }

    /// Get thumbprints of all keys in JWKS.
    ///
    /// # Arguments
    ///
    /// * `jwks` - Optional JWKS to get thumbprints from (uses cached if None)
    ///
    /// # Returns
    ///
    /// Returns a vector of key thumbprints or an error.
    pub fn thumbprints(&self, jwks: Option<&Jwks>) -> Result<Vec<String>> {
        let jwks = if let Some(jwks) = jwks {
            jwks
        } else {
            let cache = self.cache.read().unwrap();
            let cached = cache
                .as_ref()
                .ok_or_else(|| CertNodeError::Other("No JWKS available".into()))?;
            &cached.jwks
        };

        let mut thumbprints = Vec::new();
        for key in &jwks.keys {
            if let Ok(thumbprint) = crate::jwk_thumbprint(key) {
                thumbprints.push(thumbprint);
            }
            // Skip keys that can't generate thumbprints
        }

        Ok(thumbprints)
    }

    /// Validate JWKS structure and keys.
    fn validate_jwks(&self, jwks: &Jwks) -> Result<()> {
        if jwks.keys.is_empty() {
            return Err(CertNodeError::InvalidFormat("JWKS contains no keys".into()));
        }

        for (i, key) in jwks.keys.iter().enumerate() {
            self.validate_jwk(key)
                .map_err(|e| CertNodeError::InvalidFormat(format!("Key {}: {}", i, e)))?;
        }

        Ok(())
    }

    /// Validate a single JWK.
    fn validate_jwk(&self, jwk: &Jwk) -> Result<()> {
        match jwk {
            Jwk::Ec { crv, x, y, .. } => {
                if crv != "P-256" {
                    return Err(CertNodeError::UnsupportedKey(format!(
                        "Only P-256 curve supported for EC keys, got {}",
                        crv
                    )));
                }
                if x.is_empty() || y.is_empty() {
                    return Err(CertNodeError::InvalidFormat(
                        "EC key missing x or y coordinate".into(),
                    ));
                }
            }
            Jwk::Okp { crv, x, .. } => {
                if crv != "Ed25519" {
                    return Err(CertNodeError::UnsupportedKey(format!(
                        "Only Ed25519 curve supported for OKP keys, got {}",
                        crv
                    )));
                }
                if x.is_empty() {
                    return Err(CertNodeError::InvalidFormat(
                        "OKP key missing x coordinate".into(),
                    ));
                }
            }
        }

        Ok(())
    }

    /// Clear the cache.
    ///
    /// Forces the next fetch operation to retrieve fresh JWKS.
    pub fn clear_cache(&self) {
        let mut cache = self.cache.write().unwrap();
        *cache = None;
    }

    /// Check if cache contains fresh JWKS.
    pub fn has_fresh_cache(&self) -> bool {
        let cache = self.cache.read().unwrap();
        cache
            .as_ref()
            .map_or(false, |cached| cached.cached_at.elapsed() < self.ttl)
    }
}

/// Simple JWKS fetch function without caching.
///
/// This is a convenience function for one-time JWKS fetching.
/// For repeated fetches, use `JwksManager` for better performance.
///
/// # Arguments
///
/// * `url` - URL to fetch JWKS from
///
/// # Returns
///
/// Returns the JWKS or an error if fetch/parse fails.
///
/// # Examples
///
/// ```rust,no_run
/// use certnode::jwks::fetch_jwks;
///
/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
/// let jwks = fetch_jwks("https://api.certnode.io/.well-known/jwks.json").await?;
/// println!("Fetched {} keys", jwks.keys.len());
/// # Ok(())
/// # }
/// ```
#[cfg(feature = "jwks-fetch")]
#[cfg_attr(docsrs, doc(cfg(feature = "jwks-fetch")))]
pub async fn fetch_jwks(url: &str) -> Result<Jwks> {
    let client = reqwest::Client::new();

    let response = client
        .get(url)
        .timeout(Duration::from_secs(30))
        .send()
        .await?;

    if !response.status().is_success() {
        return Err(CertNodeError::NetworkError(format!(
            "HTTP {} from {}",
            response.status(),
            url
        )));
    }

    let jwks: Jwks = response.json().await?;

    // Basic validation
    if jwks.keys.is_empty() {
        return Err(CertNodeError::InvalidFormat("JWKS contains no keys".into()));
    }

    Ok(jwks)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Jwk;
    use std::time::Duration;

    #[test]
    fn test_jwks_manager_creation() {
        let manager = JwksManager::new(Duration::from_secs(300));
        assert!(!manager.has_fresh_cache());
    }

    #[test]
    fn test_set_from_object() {
        let manager = JwksManager::new(Duration::from_secs(300));

        let jwks = Jwks {
            keys: vec![Jwk::Ec {
                kty: "EC".to_string(),
                crv: "P-256".to_string(),
                x: "f83OJ3D2xF1Bg8vub9tLe1gHMzV76e8Tus9uPHvRVEU".to_string(),
                y: "x_FEzRu9m36HLN_tue659LNpXW6pCyStikYjKIWI5a0".to_string(),
                kid: Some("test-key".to_string()),
                alg: Some("ES256".to_string()),
            }],
        };

        assert!(manager.set_from_object(jwks).is_ok());
        assert!(manager.has_fresh_cache());

        let fresh = manager.get_fresh();
        assert!(fresh.is_some());
        assert_eq!(fresh.unwrap().keys.len(), 1);
    }

    #[test]
    fn test_thumbprints() {
        let manager = JwksManager::new(Duration::from_secs(300));

        let jwks = Jwks {
            keys: vec![
                Jwk::Ec {
                    kty: "EC".to_string(),
                    crv: "P-256".to_string(),
                    x: "f83OJ3D2xF1Bg8vub9tLe1gHMzV76e8Tus9uPHvRVEU".to_string(),
                    y: "x_FEzRu9m36HLN_tue659LNpXW6pCyStikYjKIWI5a0".to_string(),
                    kid: Some("test-key".to_string()),
                    alg: Some("ES256".to_string()),
                },
                Jwk::Okp {
                    kty: "OKP".to_string(),
                    crv: "Ed25519".to_string(),
                    x: "11qYAYKxCrfVS_7TyWQHOg7hcvPapiMlrwIaaPcHURo".to_string(),
                    kid: Some("ed25519-key".to_string()),
                    alg: Some("EdDSA".to_string()),
                },
            ],
        };

        let thumbprints = manager.thumbprints(Some(&jwks));
        assert!(thumbprints.is_ok());
        assert_eq!(thumbprints.unwrap().len(), 2);
    }

    #[test]
    fn test_validate_invalid_jwks() {
        let manager = JwksManager::new(Duration::from_secs(300));

        // Empty JWKS
        let empty_jwks = Jwks { keys: vec![] };
        assert!(manager.validate_jwks(&empty_jwks).is_err());

        // Invalid EC key
        let invalid_jwks = Jwks {
            keys: vec![Jwk::Ec {
                kty: "EC".to_string(),
                crv: "P-384".to_string(), // Unsupported curve
                x: "test".to_string(),
                y: "test".to_string(),
                kid: None,
                alg: None,
            }],
        };
        assert!(manager.validate_jwks(&invalid_jwks).is_err());
    }

    #[test]
    fn test_clear_cache() {
        let manager = JwksManager::new(Duration::from_secs(300));

        let jwks = Jwks {
            keys: vec![Jwk::Ec {
                kty: "EC".to_string(),
                crv: "P-256".to_string(),
                x: "f83OJ3D2xF1Bg8vub9tLe1gHMzV76e8Tus9uPHvRVEU".to_string(),
                y: "x_FEzRu9m36HLN_tue659LNpXW6pCyStikYjKIWI5a0".to_string(),
                kid: Some("test-key".to_string()),
                alg: Some("ES256".to_string()),
            }],
        };

        manager.set_from_object(jwks).unwrap();
        assert!(manager.has_fresh_cache());

        manager.clear_cache();
        assert!(!manager.has_fresh_cache());
    }
}